service/storage_proxy.cc

```diff
@@ -3845,7 +3845,22 @@ future<result<>> storage_proxy::schedule_repair(std::unordered_map<dht::token, s
     if (diffs.empty()) {
         return make_ready_future<result<>>(bo::success());
     }
-    return mutate_internal(diffs | boost::adaptors::map_values, cl, false, std::move(trace_state), std::move(permit));
+
+    auto mutations = diffs | boost::adaptors::map_values;
+
+    slogger.trace("schedule_repair cl={}", cl);
+    mlogger.trace("mutations={}", mutations);
+
+    auto type = (std::next(std::begin(mutations)) == std::end(mutations) ? db::write_type::SIMPLE : db::write_type::UNLOGGED_BATCH);
+    utils::latency_counter lc;
+    lc.start();
+
+    return mutate_prepare(mutations, cl, type, trace_state, std::move(permit), db::allow_per_partition_rate_limit::no).then(utils::result_wrap([this, cl, trace_state]
+            (storage_proxy::unique_response_handler_vector ids) mutable {
+        return mutate_begin(std::move(ids), cl, trace_state);
+    })).then_wrapped([this, p = shared_from_this(), lc, trace_state] (future<result<>> f) mutable {
+        return p->mutate_end(std::move(f), lc, get_stats(), std::move(trace_state));
+    });
```

Comment:
This causes a lot of code duplication. It's probably better to make mutate_internal accept an erm (will make it a more useful building block. @bhalevy ?

Reply:
The mutate_internal needs erm only from schedule_repair. But it's also called from many other places where erm is not needed, so all those places will just pass nullptr with no meaning.

Reply:
Can you please list those out?
I think that at least for cas, getting the erm earlier makes a lot of sense.

Reply:
paxos_response_handler::begin_and_repair_paxos
paxos_response_handler::learn_decision
storage_proxy::do_mutate
storage_proxy::replicate_counter_from_leader
storage_proxy::send_hint_to_all_replicas

Reply:
> paxos_response_handler::begin_and_repair_paxos
> paxos_response_handler::learn_decision

paxos should keep erm throughout the read-modify-write operation

> storage_proxy::do_mutate

It wouldn't hurt to get the erm in the upper layer and push it down,
or pass nullptr and let the low level fetch one when needed (when creating the result handler, right?)

> storage_proxy::replicate_counter_from_leader

We need to get a global erm in mutate_counter_on_leader_and_replicate
and use that for all shards to make sure they all use the same erm.

> storage_proxy::send_hint_to_all_replicas

hint manager should consult the erm and make sure the destination is in thew ring first and wasn't removed / decommissioned / replaced.

Reply:
Should this all come in one large PR?

Reply:
A comprehensive fix for the hints manager is out of scope, but if this can fit a minimal fixes with FIXMEs where appropriate, I'd vote to do it in a single PR.